Savanna Web app Tasks 9 - 11

22DPVEmer opened the pull request:
I added CreationIteration to AnimalState for accurate age calculation. I updated GameInstance to track offspring count and animal creation iteration. I modified GameStateDto to include Age and OffspringCount in the base animal representation. I refactored the animal details display in UIManager with dynamic health and selection. I also enhanced the login, landing page, and register design, along with improving the dashboard. Fixed the bug of tiger and zebra being able to multiply by itself. Updated UI with individual animal statistics tracking.

1. Image of updated login:
![image](https://github.com/user-attachments/assets/dc9059ac-39e6-4672-a2ee-413cf7ee43fa)

2. Image of updated Register page.
![image](https://github.com/user-attachments/assets/01f388be-2a7e-4454-b5ae-57a2976ea6d0)

3. Image of updated landing page:
![image](https://github.com/user-attachments/assets/923800e9-a5a4-45da-af90-1f301dfffcdc)

4. Image of updated dashboard page:
![image](https://github.com/user-attachments/assets/43f33270-8d09-47b2-bea3-13e8235a06d2)
5. Image of selected animal statistics and selection:
![image](https://github.com/user-attachments/assets/4532e66e-ae45-4c36-93a5-cb09259f8330)

22DPVEmer committed:
Enhance game state tracking and animal details display along with design

- Add CreationIteration to AnimalState for accurate age calculation
- Update GameInstance to track offspring count and animal creation iteration
- Modify GameStateDto to include Age and OffspringCount in base animal representation
- Refactor animal details display in UIManager with dynamic health and selection
- Improve initial animal state initialization in game management

sreeharihm commented on Savanna.Web/wwwroot/js/game/saveManager.js:
```diff
@@ -2,75 +2,72 @@
-                }
-                throw new Error(GameConstants.UI.Messages.Error.FAILED_TO_LOAD_SAVES);
-            }
+            const response = await fetch('/api/games/saved');
```
move endpoints to constant

22DPVEmer replied:
Moved it.

22DPVEmer replied:
![image](https://github.com/user-attachments/assets/a0c0a077-13b4-4bbe-b95e-244d033c67d0)
I changed the dashboard to use this image as a background. Is it any good?

sreeharihm replied:
Its better a plain rectange and background color I was mentioning about the yellow one. Make it more catchy, or you can use accenture color theme as well.

22DPVEmer replied:
![image](https://github.com/user-attachments/assets/584da6c4-cc32-4379-9390-bf680c2a7600)
I did it like this. What do you think, should I apply the background image for all pages?

22DPVEmer replied:
![image](https://github.com/user-attachments/assets/a12727fc-ad10-4375-904e-9650780888e7)
I added the search bar and basic date filtering, along with the scroll section for saves so the page is shorter.

sreeharihm reviewed (commented):
added review comments

sreeharihm commented:
In addition to above, for the improvement side, its better to have a catchy background color.
I would prefer for a savana without cells but comes under the rectangular area

22DPVEmer committed:
Refactor SaveManager to use GameConstants for API endpoint retrieval

22DPVEmer committed:
Enhance game grid styling with savanna-themed background and visual improvements

- Add savanna background image to game grid
- Improve cell hover and selection effects
- Refactor game cell and container styles for better visual hierarchy
- Update color schemes and add subtle shadows and text effects
- Remove unnecessary border and padding styles
- Improve responsiveness and visual consistency

22DPVEmer committed:
Enhanced saved games management with date filtering and search functionality

- Add dynamic filtering for saved games (today, week, month)
- Implement search functionality for saved games
- Improve saved games rendering and interaction
- Add scrollbar and styling improvements for saved games container

sreeharihm commented on Savanna.Web/wwwroot/js/game/saveManager.js:
```diff
@@ -1,76 +1,133 @@
+        if (searchInput) {
+            searchInput.addEventListener('input', (e) => {
+                const searchTerm = e.target.value.toLowerCase();
+                const filteredGames = this.filterGames(this.savedGames).filter(game => {
```
what if page refresh happens, will all the saved games should be removed from the filterGames

22DPVEmer replied:
Yes they will, is that bad?

sreeharihm replied:
yes

22DPVEmer replied:
So they should stay persistent?

sreeharihm replied:
yes, we usually implement filtering in the API side, so the refresh issues wont happen

22DPVEmer replied:
I now am using the controller for filtering that and also saving it in the URL for it to stay on refresh.

sreeharihm reviewed (commented):
usually filtering is done from the backend side.

22DPVEmer committed:
Implement server-side and client-side search functionality for saved games

- Add SearchSavedGames endpoint in GamesController to support advanced game search
- Update SaveManager with enhanced search mechanism using API call
- Add search button and URL-based search persistence
- Implement dynamic search across game save attributes
- Improve search input handling with enter key and button click events

sreeharihm commented on Savanna.Web/Controllers/GamesController.cs:
```diff
@@ -328,5 +328,64 @@ public async Task<IActionResult> DeleteSave(int saveId)
+
+                // Get the actual user ID if authenticated
+                string actualUserId = userId;
+                if (User.Identity?.IsAuthenticated == true)
```
Can you move these authentication to middleware or use authorize attribute in the controller

22DPVEmer replied:
Yes, I changed it to use authorize attribute instead for the controller and it worked fine.

sreeharihm commented on Savanna.Web/Controllers/GamesController.cs:
```diff
@@ -328,5 +328,64 @@ public async Task<IActionResult> DeleteSave(int saveId)
+                    })
+                    .Where(g => 
+                        g.SaveDate.ToString().ToLower().Contains(searchTerm) ||
+                        g.Iteration.ToString().Contains(searchTerm) ||
```
can you combine 2 where method to a single one

22DPVEmer replied:
I combined the two where clauses into a single.

22DPVEmer committed:
Refactor GamesController with authorization and improved user ID handling

sreeharihm reviewed (commented):
Added review comments

sreeharihm reviewed (approved):
Looks good!!
